megamek/src/megamek/common/units/BuildingEntity.java

```diff
@@ -0,0 +1,803 @@
+    }
+
+    @Override
+    public Building getInternalBuilding() {
+        return building;
+    }
+
+    @Override
+    public CubeCoords boardToRelative(Coords boardCoords) {
+        // Find which relative CubeCoord maps to this board coordinate
+        return relativeLayout.entrySet().stream()
+            .filter(e -> e.getValue().equals(boardCoords))
+            .map(Map.Entry::getKey)
+            .findFirst()
+            .orElse(boardCoords.toCube());
+    }
+
+    @Override
+    public Coords relativeToBoard(CubeCoords relativeCoords) {
+        // Return the board coordinate this relative CubeCoord maps to
+        return relativeLayout.get(relativeCoords);
+    }
+
+    /**
+     * Override setPosition to populate the relativeLayout map when the entity is placed.
+     * This establishes the mapping between the building's internal relative coordinates
+     * and their actual board coordinates.
+     */
+    @Override
+    public void setPosition(Coords position) {
+        super.setPosition(position);
+        updateRelativeLayout();
+    }
+
+    @Override
+    public void setPosition(Coords position, boolean gameUpdate) {
+        super.setPosition(position, gameUpdate);
+        updateRelativeLayout();
+    }
+
+    /**
+     * Updates the relativeLayout map to reflect the current building configuration.
+     * Maps each relative CubeCoord in the building to its actual board position.
+     */
+    private void updateRelativeLayout() {
+        relativeLayout.clear();
+
+        if (getPosition() == null) {
+            return;
+        }
+
+        secondaryPositions.put(0, getPosition());
+        relativeLayout.put(CubeCoords.ZERO, getPosition());
+
+        // Map each relative CubeCoord to its actual board coordinate
+        int position = 1;
+        for (CubeCoords relCoord : building.getCoordsList()) {
+            // We add the origin manually
+            if (!relCoord.equals(CubeCoords.ZERO)) {
+                // TODO: When rotation is implemented, rotate by facing before adding to entity position
+                // For now, with no rotation, convert CubeCoord to offset and add to entity position
+                CubeCoords positionCubeCoords = getPosition().toCube();
+                Coords boardCoord = positionCubeCoords.add(relCoord).toOffset();
+
+                relativeLayout.put(relCoord, boardCoord);
+                secondaryPositions.put(position, boardCoord);
+                position++;
+            }
+        }
+    }
+
+    // FIXME: IDK if this is right, just needed something to pass tests
+    private static final TechAdvancement TA_BUILDING_ENTITY = new TechAdvancement(TechBase.ALL)
+          .setAdvancement(DATE_PS, DATE_PS, DATE_PS)
+          .setTechRating(TechRating.B)
+          .setAvailability(AvailabilityValue.A, AvailabilityValue.A, AvailabilityValue.A, AvailabilityValue.A)
+          .setStaticTechLevel(SimpleTechLevel.ADVANCED);
```

Comment:
I believe this is because every Entity implements ITechnology so we have a common interface to make an entity total up its internals and tell us its overall rating and availability.

If we _have_ any TA data on mobile structures or advanced buildings, we should try to match that data, but I'm guessing it will only be important for the various subclasses - hardened buildings have existed since pre-industrial times, but within CBT we probably only care about TA for mobile structures, naval vessels, Lighter Than Air craft, etc.